File: backend/src/main/java/com/tjost/workouttracker/controller/ExerciseController.java
package com.tjost.workouttracker.controller;

import com.tjost.workouttracker.dto.ExerciseRequest;
import com.tjost.workouttracker.model.Exercise;
import com.tjost.workouttracker.service.ExerciseService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/exercises")
@RequiredArgsConstructor
public class ExerciseController {

    private final ExerciseService exerciseService;

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public Exercise createExercise(@RequestBody ExerciseRequest exercise) {
        return exerciseService.createExercise(exercise.name());
    }

    @GetMapping
    public List<Exercise> getAllExercises() {
        return exerciseService.getAllExercises();
    }

    @GetMapping("/{id}")
    public Exercise getExerciseById(@PathVariable Long id) {
        return exerciseService.getExerciseById(id);
    }

    @PutMapping("/{id}")
    public Exercise updateExercise(
            @PathVariable Long id,
            @RequestBody ExerciseRequest exercise
    ) {
        return exerciseService.updateExercise(id, exercise.name());
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void deleteExercise(@PathVariable Long id) {
        exerciseService.deleteExercise(id);
    }
}

File: backend/src/main/kotlin/workouttracker/backend/ExerciseController.kt
package workouttracker.backend

import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import workouttracker.backend.exception.ExerciseNotFoundException
import workouttracker.backend.model.Exercise
import workouttracker.backend.service.ExerciseService
import java.util.*

@RestController
@RequestMapping("/exercise")
class ExerciseController(private val exerciseService: ExerciseService) {

    @GetMapping
    fun getExercises(): ResponseEntity<List<Exercise>> {
        return ResponseEntity.ok(exerciseService.getAllExercises())
    }

    @GetMapping
    fun getExerciseById(@PathVariable id: String): ResponseEntity<Exercise> {
        val exercise = exerciseService
            .getExerciseById(id)
            .orElseThrow {ExerciseNotFoundException("Exercise with id $id not found") }

        return ResponseEntity.ok(exercise)
    }

    @GetMapping("/name/{name}")
    fun getExerciseByName(
        @PathVariable name: String
    ): ResponseEntity<Exercise> {
        val exercise = exerciseService
            .getExerciseByName(name)
            .orElseThrow { ExerciseNotFoundException("Exercise with name \"$name\" not found") }
        return ResponseEntity.ok(exercise)
    }

}